src/compiler/checker.ts

```diff
@@ -19583,6 +19597,15 @@ namespace ts {
             return isUnitType(type) || !!(type.flags & TypeFlags.TemplateLiteral);
         }
 
+        function getExactOptionalUnassignableProperties(source: Type, target: Type) {
+            return getPropertiesOfType(target)
+                .filter(targetProp => isExactOptionalPropertyMismatch(getTypeOfPropertyOfType(source, targetProp.escapedName), getTypeOfSymbol(targetProp)));
+        }
+
+        function isExactOptionalPropertyMismatch(source: Type | undefined, target: Type | undefined) {
+            return !!source && !!target && maybeTypeOfKind(source, TypeFlags.Undefined) && !!containsMissingType(target);

```

Comment:
I think I just didn’t realize how exactOptionalPropertyTypes was implemented, but it looks like this implies that we never make property types that contain _both_ `missingType` _and_ `undefinedType`. Because if we ever did, `containsMissingType(target)` would be insufficient for this check (you’d need to ensure the target _does_ contain `missingType` and _doesn’t_ contain `undefinedType`).

Reply:
Let me test that.

Reply:
Yeah I believe union reduction specifically watches out for that.